Topics/Tech_Stacks/Django_Rest.md

```diff
@@ -0,0 +1,120 @@
+ - [Django](https://www.djangoproject.com/): Web framework for Python.
+
+### What Is Django Rest Framework?
+Django Rest Framework is a popular library for building RESTful APIs with Django. It provides powerful tools for building APIs, including serialization, authentication, and permissions. You can easily create APIs that support multiple formats like JSON and XML that can be consumed by a variety of client applications. To demonstrate the power of Django Rest Framework, we will create a simple API.
+
+### Set-Up
+
+ We first set up the virtual environment and activate it using ```python3 -m venv myenv ``` and ```source myenv/bin/activate``` respectively in the project directory.
+
+ We now install Django and Django Rest Framework using the command ```pip install django djangorestframework```.
+
+ Once installed, we can start a new Django project by running: ```django-admin startproject <project-name>```. This generates the default Django folder structure including several files.
+
+ we may now create a new Django app by running: ```python manage.py startapp <app-name>```.
+ 
+ The project structure should be similar to the following:
+ 
+ ```
+ project-name/
+│
+├── app-name/
+│   ├── __init__.py
+│   ├── admin.py
+│   ├── apps.py
+│   ├── models.py
+│   ├── serializers.py
+│   ├── tests.py
+│   ├── urls.py
+│   └── views.py
+│
+├── project-name/
+│   ├── __init__.py
+│   ├── settings.py
+│   ├── urls.py
+│   └── wsgi.py
+│
+└── manage.py
+ ```
+ 
+ In order to add Django Rest Framework to your project, open the settings.py file and add 'rest_framework' to the INSTALLED_APPS list: ```INSTALLED_APPS = [..., 'rest_framework']```
+ 
+
+ ### How it Works
+ 
+ 
+ After installing and configuring Django Rest Framework, we may now create a simple API. We first create a new model to represent our API resource. To illustrate the use of Django Rest Framework, we'll create a simple model for a todo item with two fields: its title as a CharField, i.e., a string and its complete as a BooleanField, i.e., a boolean value. We first create a new file called models.py in the API app folder by adding the following code:
+
+``` {python}
+from django.db import models
+
+class TodoItem(models.Model):
+    title = models.CharField(max_length=200)
+    completed = models.BooleanField(default=False)
+```
+
+Next, we'll create a serializer to convert our model instances to JSON. We now create a new file called serializers.py in the app API folder and add the following code:
+
+``` {python}
+from rest_framework import serializers
+from .models import TodoItem
+
+class TodoItemSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = TodoItem
+        fields = ('id', 'title', 'completed')
+```
+
+Next, we'll create a view to handle incoming HTTP requests (GET, POST, PUT, PATCH and DELETE). We create a new file called views.py in the API app folder and add the following code:
+
```

Comment:
it would be very helpful if you could provide a small but complete example in how to use serialization to organize your code as it is a very important part of Django, maybe even with a small customized query selection to show that  we can easily filter out the data we want from database. I see you importing the serializer here but actually providing a small chunk of sample code would be very helpful

Reply:
I added a more detailed explanation on serialization, and gave an example with a customized query selection.